matron/src/weaver.c

```diff
@@ -2289,6 +2316,57 @@ void w_handle_crow_event(void *dev, int id) {
     l_report(lvm, l_docall(lvm, 2, 0));
 }
 
+void w_handle_serial_config(char *path, char *name, char *vendor, char *model) {
+    _push_norns_func("serial", "config");
+    lua_pushstring(lvm, vendor);
+    lua_pushstring(lvm, model);
+    l_report(lvm, l_docall(lvm, 2, 2));
+    if (lua_isnil(lvm, -2) || lua_isnil(lvm, -1)) {
+        fprintf(stderr, "no serial handler found for device %s at %s\n", name, path);
+        return;
+    }
+    if (!lua_isstring(lvm, -2)) {
+        fprintf(stderr, "serial handler id expected\n");
+        return;
+    }
+    if (!lua_istable(lvm, -1)) {
+        fprintf(stderr, "serial config table expected\n");
+        return;
+    }
+
+    dev_list_add(DEV_TYPE_SERIAL, path, strdup(name), lvm);
```

Comment:
Why is only `name` duplicated here? Where does it get freed?

Reply:
The instance of `name` being duped is owned by the event and gets freed in `event_data_free`. 

`dev_new` dupes `path` but not `name`. This works for other devices because they immediately take ownership of `name` when they are recognized. The serial devices are configured asynchronously, so I had to do things a bit differently.

Reply:
Another option would be to not free the name in `event_data_free` and leave it to `w_handle_serial_config` to either free the name on failure or pass ownership to `dev_new`. Maybe that is a bit cleaner?

Reply:
I think it's good as-is. Was originally asking as a genuine question more so than a rhetorical one -- so thanks for clearing up how it works to me :)